### PR TITLE
replace Buffers of tower services with service factories

### LINKED ISSUE #915: UX: display help when query planning on a previous generation supergraph
Apollo Router uses the Federation 2 query planner, which is capable of query planning from both **pre-Federation 2** _and_ **Federation 2** supergraphs.

Currently, when loading a previous generation supergraph into the Router, if there is something notably incorrect with the supergraph schema which wasn't detected by the previous generation query planner (present only in the JavaScript `@apollo/gateway@0.x` series) but **is** — thanks to improved validations — detected by the Federation **2** query planner, the query planner implementation (and thus, the Router, which uses the query planner) is produces an error.  That error appears similar to the following but is dependent on the actual supergraph being loaded:

```
cannot create the router: couldn't build Router Service: query planning had errors: bridge errors: UNKNOWN: Cannot find type HornStyle in subgraph unicorns
```

This error is absolutely representative of a newly found _mistake_ in the existing relationships defined in subgraphs and is worth looking at.   Often, it is a small omission from an existing subgraph (e.g., missing `@key` directive, or `@requires` field-sets which are referring to fields that don't exist) and simple to adjust!

While there isn't much that the Router **itself** can do to point the user to the site of the mistake, Federation 2's composition (which is invoked through `rover supergraph compose`) does provide substantially better diagnostics and debugging information than its predecessor.

On account of that, we may want to make it clear to those running previous generation supergraphs that there's a diagnostics path which can help unblock them.   Previous generation supergraphs can be identified through the presence of `@join` directives referencing the `join/v0.1` specification (Federation 2 supergraphs will instead reference `join/v0.2`).

This issue proposes that we display a friendly message _in addition to the error message_ with the suggestion that the user try composition with Federation 2 to find the full error message:

```
For help with this error, try re-composing your subgraphs with Federation 2 in order to benefit from its improved error messaging and diagnostics.  You can then "backport" any adjustments you discover to your existing subgraphs, without needing to fully upgrade to Federation 2 at the current time.  For details, see: https://www.apollographql.com/docs/federation/federation-2/moving-to-federation-2/#configure-your-composition-method
```

Feel free to bikeshed this message!  We can make `https://www.apollographql.com/docs/federation/federation-2/moving-to-federation-2/#configure-your-composition-method` into a `go.apollo.dev/r/` short-link to make the message appear cleaner.

### PR DESCRIPTION
**Pull request by Geal**
The router relies heavily on tower::Buffer to make services clonable, so
they can be reused for multiples requests, but this is far from the
preferred model of tower services where a new Service instance is
generated for every new connection or channel.

This commit experiments replaces the RouterService Buffer with a
RouterServiceFactory that generates a new RouterService dynamically.

This has no impact on performance and improves reliability by making
sessions fail directly on panics instead of breaking the Buffer's spawned
task.

This only affect RouterService, the other types of internal services will
be migrated in future commits.

Sidequest: fix #915

### CODE REVIEW

File: apollo-router/src/plugins/traffic_shaping/mod.rs [hunk 1]
```diff
@@ -239,12 +239,12 @@ mod test {
...
             .with_subgraph_service("products", product_service.clone());
 
-        let (router, _) = builder.build().await.expect("should build");
+        let router = builder.build().await.expect("should build").test_service();
 
         router
     }
 
-    async fn get_taffic_shaping_plugin(config: &serde_json::Value) -> Box<dyn DynPlugin> {
+    async fn get_traffic_shaping_plugin(config: &serde_json::Value) -> Box<dyn DynPlugin> {
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/plugins/traffic_shaping/mod.rs:247]: technically unrelated to that PR but I could not leave it as is :D

File: apollo-router/src/services/router_service.rs [hunk 2]
```diff
@@ -376,27 +372,95 @@ impl PluggableRouterServiceBuilder {
...
+            execution_service,
+            schema: self.schema,
+            plugins: Arc::new(self.plugins),
+            apq: APQLayer::default(),
+        })
+    }
+}
+
+#[derive(Clone)]
+pub struct MakeARouter {
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/services/router_service.rs:386]: temporary name, I'll think of more approriate ones for the entire stack (or give up and we'll have a `Router`, `RouterFactory` and `RouterFactoryFactory`)

**Review by bnjjj — APPROVED** [review]

**Comment by BrynCooke** [comment]
I've flipped backwards and forwards on this a few times and am now wondering if we should reach out to the tower project to ask them what best practice is.

Reasons for:
* linkerd does it (why?)
* `If Poll::Ready(Err(_)) is returned, the service is no longer able to service requests and the caller should discard the service instance.`
* Panics and buffers needs fixing https://github.com/tower-rs/tower/issues/455.

Reasons against:
* Various layers seem to maintain state at the service rather than at the layer and will cease to work as expected:
  * Rate limit (concurrency limit which has a global version is not the same) 
  * Hedge
  * LoadShed,
  * Balance
* The API of `tower` with use of `&mut self points towards allow state at the service level. If this was non-`mut` then it would point to putting state inside of mutexes.
* The example for how to use tower with tokio reuses a service rather than creating a new one every time: https://github.com/tower-rs/tokio-tower/blob/master/src/pipeline/server.rs#L138
* Shared docs also indicate the use of buffer is acceptable https://docs.rs/tower/latest/tower/make/struct.Shared.html
* Integration with Hyper looks to be stateful at the service layer: https://github.com/hyperium/tonic/blob/master/tonic/src/transport/service/reconnect.rs

At this point I'm not convinced either way, but think that we should get more info before merging this.

**Review by BrynCooke — COMMENTED** [review]
I'd like us to address the comments in https://github.com/apollographql/router/pull/1289#issuecomment-1172182669 before merging this.

**Comment by bnjjj** [comment]
Let me add my opinion to regarding what you said @BrynCooke 
> Reasons against:
> Various layers seem to maintain state at the service rather than at the layer and will cease to work as expected:
> Rate limit (concurrency limit which has a global version is not the same)

It's true ! I'm currently having this issue trying to implement rate limit on this implementation. The reason why I tried to implement this on this branch is because I also had issues with `Buffer`. Regarding the tower code for `Buffer` it seems that if something returns an error in `poll_ready` it won't try again, it will always return the same error. I need to double check this. I don't know what's the best strategy right now.

**Comment by BrynCooke** [comment]
> It's true ! I'm currently having this issue trying to implement rate limit on this implementation. The reason why I tried to implement this on this branch is because I also had issues with `Buffer`.

If it was just this one layer then that would be one thing, but service level state seems to be part of the design of tower.

There is definitely something that needs to be fixed, I'm just not 100% sure on how, and tower doesn't make it obvious.

My suspicion is that https://docs.rs/tower/latest/tower/reconnect/index.html may be part of the solution.

**Comment by Geal** [comment]
one of the issues here is that tower services can cover multiple patterns (like, even clients can be built with tower service, not only servers).
In this one:
>The example for how to use tower with tokio reuses a service rather than creating a new one every time: https://github.com/tower-rs/tokio-tower/blob/master/src/pipeline/server.rs#L138

It assumes it is running above a transport that sends requests one by one, so that is adapted to a single HTTP1 connection, or our usage of the query planner service, but not the router service or subgraph service.
> https://github.com/tower-rs/tokio-tower/blob/master/src/pipeline/server.rs#L11-L13
/// This type provides an implementation of a Tower
/// [`Service`](https://docs.rs/tokio-service/0.1/tokio_service/trait.Service.html) on top of a
/// request-at-a-time protocol transport.

That would work well in an application that needs a long running session with an external service but it is less adapted to a server.
On the server side, if we look at hyper, the [serve()](https://docs.rs/hyper/0.14.19/hyper/server/struct.Builder.html#method.serve) method expect a `MakeServiceRef` as argument. The type is private but if you follow around the code, it's creating a new service instance on each new HTTP session (each new connection in HTTP 1, each new multiplexed stream in HTTP/2). That service instance can be called multiple times in a row.

Some important points here:
- creating a new service instance per session can be done by creating a service from scratch, applying layers, etc, but also by cloning an existing service (if it implements `Clone`, which `Buffer` gives us)
- moving from one shared service to creating service instances on the fly does not change the state management that much: if there's some shared immutable state it will be cloned in every service instance(cheap with an `Arc`), if there's some shared mutable state it will be cloned, moved into the future returned by `call()` and access will be synchronized there
- but the big difference is that we can now decide if we want global (server level) or local (session level) state management. We might want to have global rate limiting for all the sessions, but also have per session rate limiting

So I think the point here is that instead of using a `Buffer` everywhere, we can explicitely decide which model we will use, per service. For the `RouterService` and `ExecutionService` it makes more sense to create one service per session, but for the `QueryPlanner` we would better synchronize and queue access. For subgraph services I do not know yet :smile:

**Comment by Geal** [comment]
yes, it looks like they fit different models, but the documentation does not make it clear which one is preferred for which use case.

I think for outbound it's unclear because there are shared and not shared parts:
- cache and query deduplication are shared
- service discovery, load balancing and circuit breaking would be shared too
- but retry and reconnect would not: ideally one service per backend instance, but then how do we handle multiple connections to one subgraph (iirc there's a connection pool handled by reqwest or hyper transparently)

**Comment by BrynCooke** [comment]
Suggest we do a little more investigation to ensure we've got it right. Not too long as we don't want to let the PR rot, but aim for clarity next week.

**Pull request merged**